projects/libsodium/secretbox_easy_fuzzer.cc

```diff
@@ -0,0 +1,23 @@
+#include <sodium.h>
+
+#include "fake_random.h"
+
+extern "C" int LLVMFuzzerTestOneInput(const unsigned char *data, size_t size) {
+  setup_sodium_w_deterministic_random();
+
+  unsigned char key[crypto_secretbox_KEYBYTES];
+  unsigned char nonce[crypto_secretbox_NONCEBYTES];
+
+  crypto_secretbox_keygen(key);
```

Comment:
Given that I'm using the deterministic random bytes generator this shouldn't change. With that in mind, would it make more sense just to hard code the bytes to the key and the nonce?

Reply:
Assuming my understanding below is correct, I think it shouldn't make much of a difference.

If I have the following code:
`
LLVMFuzzerTestOneInput({0, 1, 2}, 3);
LLVMFuzzerTestOneInput({0, 1, 2}, 3);
`
Will `key` in the first call to `LLVMFuzzerTestOneInput` be the same as `key` `LLVMFuzzerTestOneInput` in the second call? Will `nonce` be the same across calls?

My understanding is that `key` and `nonce` will be the same on every call. So hardcoding them would just be more straightforward but wouldn't change much.

I should point out that if `key` and `nonce` can affect what code in your program is actually run, then a better approach is letting them change in a way determined by the input to `LLVMFuzzerTestOneInput`.

Reply:
> If I have the following code:
LLVMFuzzerTestOneInput({0, 1, 2}, 3); LLVMFuzzerTestOneInput({0, 1, 2}, 3);
Will key in the first call to LLVMFuzzerTestOneInput be the same as key LLVMFuzzerTestOneInput in the second call? Will nonce be the same across calls?

Yes, with the current implemenation of the `fake_random_buffer`, there would be no change between calls. The implementation should be entirely deterministic with the only parameter being the size of the buffer. But, given the same buffer size, the same bytes will be returned. In practice this would mean that same keys/nonces would be returned _every_ time by the fuzzer because the seed doesn't change and their size is fixed. I don't think that this is an improvement over hard-coding the keys, but I'll address that below :)

> I should point out that if key and nonce can affect what code in your program is actually run, then a better approach is letting them change in a way determined by the input to LLVMFuzzerTestOneInput.

Excellent point - it certainly seems like `key`, `nonce`, and the like could drive changes in the code. I think this is an argument in favor of pulling `fake_random.h` back in and consuming it. I intially tried to use a bit of global state to get the user data into the seed but found that difficult and a bit confusing. The only other option I thought of was to use the size of the buffer to drive the random generator, but to keep its seed constant. The problem with this approach is that given the same buffer size, I'll always get the same output.

Overall, it sounds like I *should* use the deterministic random bytes generator but that I need to find a way to make it appear a bit more random to be of value (while still being entirely deterministic given the inputs to the fuzzer.